src/client/datascience/interactive-window/interactiveWindow.ts

```diff
@@ -275,34 +267,36 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
                 ).controller.onDidChangeSelectedNotebooks(
                     (selectedEvent: { notebook: NotebookDocument; selected: boolean }) => {
                         // Controller was deselected for this InteractiveWindow's NotebookDocument
-                        if (selectedEvent.selected === false && selectedEvent.notebook === notebookDocument) {
-                            this.kernelLoadPromise = undefined;
-                            this.kernel = undefined;
+                        if (
+                            selectedEvent.selected === false &&
+                            selectedEvent.notebook.uri.toString() === currentNotebookUri
+                        ) {
                             this.notebookController = undefined;
+                            this.executionPromise = undefined;
                             controllerChangeListener.dispose();
                         }
                     },
                     this,
                     this.disposables
                 );
 
-                this.registerKernel(e.notebook, e.controller);
+                this.notebookController = e.controller;
+                this.kernelReadyPromise().ignoreErrors();
                 this.initialControllerSelected.resolve();
             },
             this,
             this.disposables
         );
     }
 
-    private registerKernel(notebookDocument: NotebookDocument, controller: VSCodeNotebookController) {
-        const kernel = this.kernelProvider.getOrCreate(notebookDocument, {
-            metadata: controller.connection,
-            controller: controller.controller,
+    private async kernelReadyPromise() {
+        const kernel = this.kernelProvider.getOrCreate(this.notebookDocument!, {
+            metadata: this.notebookController!.connection,
+            controller: this.notebookController!.controller,
             resourceUri: this.owner
         });
-        this.kernelLoadPromise = kernel?.start({ disableUI: false, document: notebookDocument });
-        this.kernel = kernel;
-        this.notebookController = controller;
+        await kernel.start({ disableUI: false, document: this.notebookDocument! });
```

Comment:
Every time you ask for this you'll be 'starting' the kernel? DOes the kernel handle this case and skip starting more than once?

If it does, can you put a comment here to say that? (so somebody in the future doesn't think this is broken)

Reply:
Yeah the IKernel caches the notebookPromise, so calling start again will be a noop. https://github.com/microsoft/vscode-jupyter/blob/2e1c221cdc3ec1128d55886711677f77d699e552/src/client/datascience/jupyter/kernels/kernel.ts#L231 I'll add a comment

Reply:
I'm a tiny bit confused on this function as well. With the name I'd expect it to just be returning a promise to await for something. But it's starting up a kernel and also (if I'm reading this right) might be returning different promises if the controller changes.